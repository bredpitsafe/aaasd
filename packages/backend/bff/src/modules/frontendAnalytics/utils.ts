import type { TBasicLogger } from '@backend/utils/src/logger';
import { assertNever } from '@common/utils';

import { ELogLevel } from './defs';

export const toLoggerMethod = (
    frontendLogLevel: ELogLevel,
): keyof Omit<TBasicLogger, 'createChildLogger'> => {
    switch (frontendLogLevel) {
        case ELogLevel.Info:
            return 'info';
        case ELogLevel.Warn:
            return 'warn';
        case ELogLevel.Error:
            return 'error';
        case ELogLevel.Debug:
            return 'debug';
        default:
            assertNever(frontendLogLevel);
    }
};
