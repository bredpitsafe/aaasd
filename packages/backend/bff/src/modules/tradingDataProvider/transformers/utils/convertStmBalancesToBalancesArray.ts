import type { TStmBalance as TStmBalanceSDK } from '@grpc-schemas/trading_data_provider-api-sdk/services/trading_data_provider/v1/stm_api.js';
import { omit } from 'lodash-es';

import type { TStmBalance } from '../../schemas/SubscribeToStmBalances.schema.ts';

export const convertStmBalancesToBalancesArray = (balances?: TStmBalanceSDK[]): TStmBalance[] =>
    balances?.map((update) => omit(update, 'position')) ?? [];
