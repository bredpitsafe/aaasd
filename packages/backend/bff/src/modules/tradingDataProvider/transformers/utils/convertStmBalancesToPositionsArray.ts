import type { TStmBalance } from '@grpc-schemas/trading_data_provider-api-sdk/services/trading_data_provider/v1/stm_api.js';
import { omit } from 'lodash-es';

import type { TStmPosition } from '../../schemas/SubscribeToStmPositions.schema.ts';

export const convertStmBalancesToPositionsArray = (balances?: TStmBalance[]): TStmPosition[] =>
    balances?.map((update) => omit(update, 'balances')) ?? [];
