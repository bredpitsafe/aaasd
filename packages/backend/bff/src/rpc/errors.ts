import { isObject, isString } from 'lodash-es';

import type { TRpcError } from '../def/rpc.ts';
import { ERpcErrorCode } from '../def/rpc.ts';
import { GrpcClientError } from '../transport/grpc/client/error.ts';

export class RpcError extends Error implements TRpcError {
    constructor(
        public code: ERpcErrorCode,
        public description: string,
        public args: Record<string, unknown> = {},
    ) {
        super(description);
        Object.setPrototypeOf(this, RpcError.prototype);
        Error.captureStackTrace(this, RpcError);
    }

    public toJSON() {
        return {
            // Convert RPC code from value to key.
            // e.g. 16 => UNAUTHENTICATED
            code: Object.entries(ERpcErrorCode).find(([, value]) => value === this.code)?.[0],
            description: this.description,
            args: this.args,
        };
    }
}

export function toRpcError(err: unknown): RpcError {
    if (err instanceof RpcError) {
        return err;
    }
    if (err instanceof GrpcClientError) {
        return new RpcError(err.code as unknown as ERpcErrorCode, err.message, {
            originalError: err,
        });
    }
    if (isObject(err) && 'message' in err && isString(err.message)) {
        return internalProcessingError(err.message);
    }
    return internalProcessingError(err);
}

/**
 * @public
 */
export function internalProcessingError(originalError: unknown) {
    return new RpcError(
        ERpcErrorCode.INTERNAL,
        'Request processing error. Internal error occurred while processing the request. Please file a bug report and provide client logs.',
        { originalError },
    );
}

/**
 * @public
 */
export function requestTimeoutError(timeoutMs: number) {
    return new RpcError(
        ERpcErrorCode.DEADLINE_EXCEEDED,
        `The request failed to receive a response within the specified timeout of ${timeoutMs}ms.`,
        { timeoutMs },
    );
}

/**
 * @public
 */
export function requestParseError(originalError: string) {
    return new RpcError(
        ERpcErrorCode.INVALID_ARGUMENT,
        'Failed to parse request body. Please file a bug report and provide client logs.',
        { originalError },
    );
}

/**
 * @public
 */
export function validationFailError(message: string) {
    return new RpcError(ERpcErrorCode.INVALID_ARGUMENT, `request validation error: ${message}`);
}

/**
 * @public
 */
export function notAuthenticatedError() {
    return new RpcError(
        ERpcErrorCode.UNAUTHENTICATED,
        'Request authentication error: the request is not authenticated.',
    );
}

/**
 * @public
 */
export function tokenVerificationError(originalError: unknown) {
    return new RpcError(
        ERpcErrorCode.UNAUTHENTICATED,
        'Request authentication error: the request is not authenticated.',
        { originalError },
    );
}
