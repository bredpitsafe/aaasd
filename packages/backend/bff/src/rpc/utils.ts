import { isNil } from 'lodash-es';

import type { TRpcResponse, TRpcRouteName } from '../def/rpc.ts';
import type { TRoutesConfig } from './def.ts';

// Needed for correct typing
export function createRpcRoutes<T extends TRpcRouteName>(
    routesConfig: TRoutesConfig<T>,
): TRoutesConfig<T> {
    return routesConfig;
}

export function isErrorResponse(response: TRpcResponse): boolean {
    return !('payload' in response) && !isNil(response?.error);
}
