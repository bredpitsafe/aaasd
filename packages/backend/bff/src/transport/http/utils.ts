import type { EHttpRouteName, THttpRoute } from './def.ts';

export function createHttpRoutes<T extends EHttpRouteName>(
    routesConfig: Record<T, THttpRoute>,
): Record<T, THttpRoute> {
    return routesConfig;
}
