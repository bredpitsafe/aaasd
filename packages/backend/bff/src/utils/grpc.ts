import type { Opaque } from '@common/types';
import Long from 'long';

/**
 * @public
 */
export type TDuration = {
    seconds: number;
    nanos: number;
};

export type TDurationString = Opaque<'DurationString', string>;

/**
 * @public
 */
export function durationToString(duration: TDuration): TDurationString {
    const seconds = duration.seconds;
    const nanos = duration.nanos;
    return Long.fromNumber(seconds)
        .multiply(1_000_000_000)
        .add(nanos)
        .div(1_000)
        .toString() as TDurationString;
}

/**
 * @public
 */
export function stringToDuration(durationString: TDurationString): TDuration {
    const value = durationString.replace(/ms$/, '');
    const seconds = Long.fromString(value).div(1_000_000_000).toNumber();
    const nanos = Long.fromString(value).mod(1_000_000_000).toNumber();
    return {
        seconds,
        nanos,
    };
}
