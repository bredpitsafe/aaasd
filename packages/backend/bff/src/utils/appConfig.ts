import { EAppEnv, EApplicationName } from '@common/types';
import { assert } from '@common/utils';
import configPackage from 'config';
import { isEmpty, isNil } from 'lodash-es';

import type { TAppConfig } from '../def/appConfig.ts';
import { EGrpcClientName } from '../def/grpcClients.ts';
import { EStageCategory, EStageEnv } from '../def/stages.ts';

export const appConfig = configPackage.util.toObject() as unknown as TAppConfig;

export const validateAppConfig = (): void => {
    const { service, stages = {}, oauth } = appConfig;
    const listStages = Object.entries(stages);

    // Check that service has OAuth URL set
    assert(!isEmpty(oauth.url), 'OAuth URL is not set');

    // Check that service actually has correct env configured
    assert(
        Object.values(EAppEnv).includes(service.env),
        `incorrect service env '${service.env}', should be one of: ${Object.values(EAppEnv)}`,
    );

    // Check that service contains any stages at all (if not, it may lack proper NODE_ENV or NODE_CONFIG_ENV variable set
    assert(!isEmpty(listStages), 'service configuration contains no stages');

    // Check that service contains only allowed stages (depending on configured env)

    listStages.forEach(([name, stage]) => {
        // Check that stage.env is set and contains valid value
        assert(
            !isNil(stage.env) && Object.values(EStageEnv).includes(stage.env),
            `stage ${name} has incorrect env '${stage.env}', should be one of: ${Object.values(
                EStageEnv,
            )}`,
        );

        // Check that stage.category is set and contains valid value
        assert(
            !isNil(stage.category) && Object.values(EStageCategory).includes(stage.category),
            `stage ${name} has incorrect category '${
                stage.category
            }', should be one of: ${Object.values(EStageCategory)}`,
        );

        // Check that stage.clientApps is set when stage is `platform` or `client`
        if (
            stage.category === EStageCategory.Platform ||
            stage.category === EStageCategory.Client
        ) {
            assert(Array.isArray(stage.clientApps), `stage '${name}' has no client apps set`);
        }

        // Check that stage.clientApps include only valid app names
        const validAppNames = Object.values(EApplicationName);
        stage.clientApps?.forEach((clientApp) => {
            assert(
                validAppNames.includes(clientApp),
                `stage '${name}' has invalid client app name '${clientApp}'`,
            );
        });

        // Check that stage.grpcClients include only valid service names
        const validGrpcClientNames = Object.values(EGrpcClientName);
        stage.grpcClients?.forEach((grpcClientName) => {
            assert(
                validGrpcClientNames.includes(grpcClientName),
                `stage '${name}' has invalid service name '${grpcClientName}'`,
            );
        });
    });
};
