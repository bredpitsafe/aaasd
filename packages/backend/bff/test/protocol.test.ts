import { generateCorrelationId } from '@backend/utils/src/correlationId.ts';
import { generateTraceId } from '@common/utils';
import { wait } from '@common/utils/src/wait.ts';
import { omit } from 'lodash-es';
import fetch from 'node-fetch';
import { firstValueFrom, fromEvent, fromEventPattern } from 'rxjs';
import { afterAll, describe, expect, it, vi } from 'vitest';
import WebSocket from 'ws';

import type { TStageName } from '../src/def/stages.ts';
import { RpcError } from '../src/rpc/errors.ts';
import { BASE_URL, Client, users, WS_ADDRESS } from './client/index.ts';

describe('Protocol', () => {
    const client = new Client();

    afterAll(() => {
        client.close();
    });

    it('healthcheck', async () => {
        const response = await fetch(`http://${BASE_URL}/healthcheck`);
        expect(await response.text()).toMatchSnapshot();
    });

    it('metrics', async () => {
        const response = await fetch(`http://${BASE_URL}/metrics`);
        expect(response.status).toBe(200);
    });

    it('ping', async () => {
        const response = await client.request({ type: 'Ping' });
        expect(response.state).toBe('Done');
        expect(response.payload?.type).toBe('Pong');
    });

    it('client heartbeat', async () => {
        const obs$ = client.send({ type: 'Heartbeat' });
        const subscriber = vi.fn();
        const subscription = obs$.subscribe(subscriber);
        await wait();

        expect(subscriber.mock.calls).toHaveLength(0);
        subscription.unsubscribe();
    });

    it('server heartbeat', async () => {
        const correlationId = generateCorrelationId();
        const obs$ = client.send({ type: 'ServerHeartbeat' }, { correlationId });
        const subscriber = vi.fn();
        const subscription = obs$.subscribe(subscriber);
        await wait(500);

        expect(subscriber.mock.calls.length).toBeGreaterThan(0);
        expect(subscriber.mock.calls[0][0]).toMatchObject({
            state: 'InProgress',
            payload: {
                type: 'Heartbeat',
            },
        });

        subscription.unsubscribe();
        await client.request({ type: 'Unsubscribe' }, { correlationId });
    });

    it('traceId', async () => {
        const traceId = generateTraceId();
        const response = await client.request({ type: 'Ping' }, { traceId });
        expect(response.traceId).toBe(traceId);
    });

    it('correlationId', async () => {
        const correlationId = generateCorrelationId();
        const response = await client.request({ type: 'Ping' }, { correlationId });
        expect(response.correlationId).toBe(correlationId);
    });

    it('unknown message type', async () => {
        // Intentionally send incorrect message, therefore ignore TS error
        // @ts-ignore
        const response = await client.request({ type: 'UNKNOWN_MESSAGE' });
        expect(response.error).toMatchSnapshot();
    });

    it('missing traceId', async () => {
        const response = await client.request({ type: 'Ping' }, { traceId: undefined });
        expect(response.error).toMatchSnapshot();
    });

    it('missing timestamp', async () => {
        const response = await client.request({ type: 'Ping' }, { timestamp: undefined });
        expect(response.error).toMatchSnapshot();
    });

    it('missing payload', async () => {
        // Intentionally send incorrect payload, therefore ignore TS error
        // @ts-ignore
        const response = await client.request(undefined, { timestamp: undefined });
        expect(response.error).toMatchSnapshot();
    });

    it('missing correlationId', async () => {
        try {
            await client.request({ type: 'Ping' }, { correlationId: undefined });
            // Client is not able to resolve a request. It'll throw an error upon encountering unknown response (with non-matching correlationId).
        } catch (err) {
            if (err instanceof RpcError && 'response' in err.args) {
                expect(
                    omit(err.args.response as object, ['traceId', 'correlationId', 'timestamp']),
                ).toMatchSnapshot();
                return;
            }
            throw err;
        }
    });

    it('non-JSON payload', async () => {
        const ws = new WebSocket(WS_ADDRESS);
        await firstValueFrom(fromEvent(ws, 'open'));
        const handleMessage = (handler: Function) =>
            ws.on('message', (message) => {
                const response = JSON.parse(message.toString());
                handler(response);
            });
        const promise = firstValueFrom<{ state: string; error: object }>(
            fromEventPattern(handleMessage),
        );
        ws.send('NOT_A_JSON');
        const message = await promise;
        expect(message.state).toBe('Done');
        expect(message.error).toMatchSnapshot();
    });

    it('simulate internal error', async () => {
        const response = await client.request({ type: 'Ping', simulateInternalError: true });

        expect(response.error).toMatchSnapshot();
    });

    it('simulate subscription timeout', async () => {
        const response = await client.request({ type: 'Ping', simulateTimeout: true });

        expect(response.error).toMatchSnapshot();
    }, 10_000);

    it('missing `requestStage` parameter', async () => {
        const authenticatedClient = new Client({ authenticate: true, as: users.user1.username });
        // Intentionally ignore TS error (missing field) because that what we're testing here
        // @ts-ignore
        const response = await authenticatedClient.request({
            type: 'SubscribeToStmBalances',
        });

        expect(response.error).toMatchSnapshot();
    });

    it('incorrect `requestStage` parameter (completely invalid)', async () => {
        const authenticatedClient = new Client({ authenticate: true, as: users.user1.username });
        const response = await authenticatedClient.request({
            type: 'SubscribeToStmBalances',
            requestStage: 'some_fake_stage' as TStageName,
        });

        expect(response.error).toMatchSnapshot();
    });

    it('incorrect `requestStage` parameter (stage does not support service)', async () => {
        const authenticatedClient = new Client({ authenticate: true, as: users.user1.username });
        const response = await authenticatedClient.request({
            type: 'SubscribeToStmBalances',
            requestStage: 'authz' as TStageName,
        });

        expect(response.error).toMatchSnapshot();
    });
});
