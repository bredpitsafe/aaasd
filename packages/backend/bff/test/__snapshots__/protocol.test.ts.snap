// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Protocol > healthcheck 1`] = `"Ok."`;

exports[`Protocol > incorrect \`requestStage\` parameter (completely invalid) 1`] = `
{
  "args": {},
  "code": "INTERNAL",
  "description": "missing stage configuration for 'some_fake_stage', check service configuration",
}
`;

exports[`Protocol > incorrect \`requestStage\` parameter (stage does not support service) 1`] = `
{
  "args": {},
  "code": "INTERNAL",
  "description": "GRPC Client 'StmV1' is not enabled for stage 'authz', check service configuration.",
}
`;

exports[`Protocol > missing \`requestStage\` parameter 1`] = `
{
  "args": {},
  "code": "INTERNAL",
  "description": "missing stage, check request parameters or service configuration",
}
`;

exports[`Protocol > missing correlationId 1`] = `
{
  "error": {
    "args": {},
    "code": "INVALID_ARGUMENT",
    "description": "request validation error: missing correlationId",
  },
  "state": "Done",
}
`;

exports[`Protocol > missing payload 1`] = `
{
  "args": {},
  "code": "INVALID_ARGUMENT",
  "description": "request validation error: missing payload",
}
`;

exports[`Protocol > missing timestamp 1`] = `
{
  "args": {},
  "code": "INVALID_ARGUMENT",
  "description": "request validation error: missing timestamp",
}
`;

exports[`Protocol > missing traceId 1`] = `
{
  "args": {},
  "code": "INVALID_ARGUMENT",
  "description": "request validation error: missing traceId",
}
`;

exports[`Protocol > non-JSON payload 1`] = `
{
  "args": {
    "originalError": "Unexpected token 'N', "NOT_A_JSON" is not valid JSON",
  },
  "code": "INVALID_ARGUMENT",
  "description": "Failed to parse request body. Please file a bug report and provide client logs.",
}
`;

exports[`Protocol > simulate internal error 1`] = `
{
  "args": {
    "originalError": "Simulated Internal Error",
  },
  "code": "INTERNAL",
  "description": "Request processing error. Internal error occurred while processing the request. Please file a bug report and provide client logs.",
}
`;

exports[`Protocol > simulate subscription timeout 1`] = `
{
  "args": {
    "timeoutMs": 300,
  },
  "code": "DEADLINE_EXCEEDED",
  "description": "The request failed to receive a response within the specified timeout of 300ms.",
}
`;

exports[`Protocol > unknown message type 1`] = `
{
  "args": {},
  "code": "INVALID_ARGUMENT",
  "description": "request validation error: missing request route: invalid 'type' field value",
}
`;
